File: lion/parse.py
import re

class Token:
    def __init__(self, type_, value, line, col):
        self.type = type_
        self.value = value
        self.line = line
        self.col = col

    def __repr__(self):
        return f"Token({self.type}, {repr(self.value)})"

class Lexer:
    def __init__(self, text):
        self.text = text
        self.tokens = []
        self.line = 1
        self.indent_stack = [0]
        
    def tokenize(self):
        lines = self.text.split('\n')
        for i, line in enumerate(lines):
            self.line = i + 1
            
            # Handle comments
            if '#' in line:
                line = line.split('#')[0]
                
            stripped = line.lstrip()
            if not stripped:
                continue
                
            indent = len(line) - len(stripped)
            
            if indent > self.indent_stack[-1]:
                self.indent_stack.append(indent)
                self.tokens.append(Token('INDENT', indent, self.line, 0))
            elif indent < self.indent_stack[-1]:
                while self.indent_stack[-1] > indent:
                    self.indent_stack.pop()
                    self.tokens.append(Token('DEDENT', self.indent_stack[-1], self.line, 0))
            
            self._tokenize_line(stripped)
            self.tokens.append(Token('NEWLINE', '\n', self.line, len(line)))
            
        while len(self.indent_stack) > 1:
            self.indent_stack.pop()
            self.tokens.append(Token('DEDENT', 0, self.line, 0))
            
        self.tokens.append(Token('EOF', None, self.line, 0))
        return self.tokens

    def _tokenize_line(self, line):
        pos = 0
        while pos < len(line):
            c = line[pos]
            start_col = pos
            if c.isspace():
                pos += 1
                continue
            if c in '()[]=':
                self.tokens.append(Token(c, c, self.line, start_col))
                pos += 1
                continue
            if c == '\"':
                end = pos + 1
                while end < len(line) and not (line[end] == '\"' and line[end-1] != '\\'):
                    end += 1
                val = line[pos+1:end].replace('\\\"', '\"')
                self.tokens.append(Token('STRING', val, self.line, start_col + pos))
                pos = end + 1
                continue
                
            match = re.match(r'[^\s()\[\]=]+', line[pos:])
            if match:
                val = match.group(0)
                if val in ('True', 'true'):
                    self.tokens.append(Token('BOOLEAN', True, self.line, start_col + pos))
                elif val in ('False', 'false'):
                    self.tokens.append(Token('BOOLEAN', False, self.line, start_col + pos))
                elif val in ('None', 'null'):
                    self.tokens.append(Token('NONE', None, self.line, start_col + pos))
                else:
                    try:
                        if '.' in val:
                            self.tokens.append(Token('NUMBER', float(val), self.line, start_col + pos))
                        else:
                            self.tokens.append(Token('NUMBER', int(val), self.line, start_col + pos))
                    except ValueError:
                        self.tokens.append(Token('IDENTIFIER', val, self.line, start_col + pos))
                pos += len(val)
            else:
                raise ValueError(f"Unexpected character '{c}' at line {self.line}")

class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0
        
    def peek(self):
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return self.tokens[-1]
        
    def consume(self, expected_type=None):
        tok = self.peek()
        if expected_type and tok.type != expected_type:
            raise ValueError(f"Expected {expected_type}, got {tok.type} at line {tok.line}")
        self.pos += 1
        return tok

    def _skip_newlines(self):
        while self.peek().type == 'NEWLINE':
            self.consume('NEWLINE')

    def parse(self):
        self._skip_newlines()
        tok1 = self.peek()
        if tok1.type == '[':
            return self.parse_list()
        return self.parse_dict()

    def parse_dict(self, end_tokens=('EOF', 'DEDENT', ')', ']')):
        res = {}
        while self.peek().type not in end_tokens:
            self._skip_newlines()
            if self.peek().type in end_tokens:
                break
            key_tok = self.consume('IDENTIFIER')
            key = key_tok.value
            
            array_depth = 0
            while True:
                if self.peek().type == '[':
                    if self.pos + 1 < len(self.tokens) and self.tokens[self.pos+1].type == ']':
                        self.consume('[')
                        self.consume(']')
                        array_depth += 1
                    else:
                        break
                else:
                    break

            if self.peek().type == 'NEWLINE':
                if self.pos + 1 < len(self.tokens) and self.tokens[self.pos+1].type == 'INDENT':
                    res[key] = self.parse_tabular_block(array_depth)
                    continue
                else:
                    if array_depth > 0:
                        res[key] = []
                    else:
                        res[key] = {}
                    continue

            if self.peek().type == '=':
                self.consume('=')
                res[key] = self.parse_value()
            elif self.peek().type == '[':
                res[key] = self.parse_list()
            elif self.peek().type == '(':
                res[key] = self.parse_inline_group()
            else:
                res[key] = self.parse_value()
                
        return res

    def parse_tabular_block(self, array_depth):
        self.consume('NEWLINE')
        self.consume('INDENT')
        
        if array_depth >= 2:
            res = []
            while self.peek().type not in ('DEDENT', 'EOF'):
                self._skip_newlines()
                if self.peek().type == 'DEDENT':
                    break
                row = self.parse_tuple_row()
                res.append(row)
            self.consume('DEDENT')
            return res
            
        headers = []
        while self.peek().type not in ('NEWLINE', 'EOF'):
            col_key = self.consume('IDENTIFIER').value
            if self.peek().type == '(':
                self.consume('(')
                sub_keys = []
                while self.peek().type not in (')', 'NEWLINE', 'EOF'):
                    sub_keys.append(self.consume('IDENTIFIER').value)
                self.consume(')')
                headers.append((col_key, sub_keys))
            else:
                headers.append(col_key)
                
        self.consume('NEWLINE')
        
        res = []
        while self.peek().type not in ('DEDENT', 'EOF'):
            self._skip_newlines()
            if self.peek().type == 'DEDENT':
                break
            
            row_dict = {}
            for header in headers:
                if isinstance(header, tuple):
                    main_k, sub_k = header
                    val = self.parse_value()
                    if isinstance(val, (tuple, list)):
                        row_dict[main_k] = dict(zip(sub_k, val))
                    else:
                        raise ValueError(f"Expected tuple/list for {main_k}")
                else:
                    row_dict[header] = self.parse_value()
                    
            res.append(row_dict)
            
        self.consume('DEDENT')
        return res if array_depth > 0 else res[0]

    def parse_list(self):
        self.consume('[')
        res = []
        while self.peek().type != ']':
            while self.peek().type in ('NEWLINE', 'INDENT', 'DEDENT'):
                self.consume()
            if self.peek().type == ']':
                break
            res.append(self.parse_value())
        self.consume(']')
        return res

    def parse_tuple_row(self):
        row = []
        while self.peek().type not in ('NEWLINE', 'EOF', 'DEDENT'):
            row.append(self.parse_value())
            if self.peek().type == ',':
                self.consume(',')
        return row

    def parse_inline_group(self):
        self.consume('(')
        is_dict = False
        temp = self.pos
        while temp < len(self.tokens) and self.tokens[temp].type != ')':
            if self.tokens[temp].type == '=':
                is_dict = True
                break
            temp += 1
            
        if is_dict:
            res = self.parse_dict(end_tokens=(')', 'NEWLINE', 'EOF'))
            self.consume(')')
        else:
            res = []
            while self.peek().type != ')':
                res.append(self.parse_value())
            self.consume(')')
            res = tuple(res)
        return res

    def parse_value(self):
        tok = self.peek()
        if tok.type == '[':
            return self.parse_list()
        if tok.type == '(':
            return self.parse_inline_group()
        if tok.type in ('STRING', 'NUMBER', 'BOOLEAN', 'NONE', 'IDENTIFIER'):
            self.consume()
            return tok.value
        raise ValueError(f"Unexpected token {tok.type} in parse_value at line {tok.line}")

def loads(text: str):
    lexer = Lexer(text)
    tokens = lexer.tokenize()
    parser = Parser(tokens)
    return parser.parse()

File: parsers/javascript/src/parse.ts
export class Token {
    constructor(
        public type: string,
        public value: any,
        public line: number,
        public col: number
    ) {}
}

export class Lexer {
    private tokens: Token[] = [];
    private line: number = 1;
    private indentStack: number[] = [0];

    constructor(private text: string) {}

    public tokenize(): Token[] {
        const lines = this.text.split('\n');
        for (let i = 0; i < lines.length; i++) {
            this.line = i + 1;
            let lineStr = lines[i].replace(/\r$/, '');

            if (lineStr.includes('#')) {
                lineStr = lineStr.split('#')[0];
            }

            const stripped = lineStr.trimStart();
            if (!stripped) {
                this.tokens.push(new Token('NEWLINE', '\n', this.line, 0));
                continue;
            }

            const indent = lineStr.length - stripped.length;

            if (indent > this.indentStack[this.indentStack.length - 1]) {
                this.indentStack.push(indent);
                this.tokens.push(new Token('INDENT', indent, this.line, 0));
            } else if (indent < this.indentStack[this.indentStack.length - 1]) {
                while (this.indentStack[this.indentStack.length - 1] > indent) {
                    this.indentStack.pop();
                    this.tokens.push(new Token('DEDENT', this.indentStack[this.indentStack.length - 1], this.line, 0));
                }
            }

            this._tokenizeLine(stripped);
            this.tokens.push(new Token('NEWLINE', '\n', this.line, lineStr.length));
        }

        while (this.indentStack.length > 1) {
            this.indentStack.pop();
            this.tokens.push(new Token('DEDENT', 0, this.line, 0));
        }

        this.tokens.push(new Token('EOF', null, this.line, 0));
        return this.tokens;
    }

    private _tokenizeLine(lineStr: string) {
        let pos = 0;
        while (pos < lineStr.length) {
            const c = lineStr[pos];
            const startCol = pos;

            if (/\s/.test(c)) {
                pos++;
                continue;
            }

            if ('()[]={}'.includes(c)) {
                this.tokens.push(new Token(c, c, this.line, startCol));
                pos++;
                continue;
            }

            if (c === '"') {
                let end = pos + 1;
                while (end < lineStr.length && !(lineStr[end] === '"' && lineStr[end - 1] !== '\\')) {
                    end++;
                }
                const val = lineStr.substring(pos + 1, end).replace(/\\"/g, '"');
                this.tokens.push(new Token('STRING', val, this.line, startCol + pos));
                pos = end + 1;
                continue;
            }

            const match = lineStr.substring(pos).match(/^[^\s()[\]={}\"]+/);
            if (match) {
                const val = match[0];
                if (val === 'True' || val === 'true') {
                    this.tokens.push(new Token('BOOLEAN', true, this.line, startCol + pos));
                } else if (val === 'False' || val === 'false') {
                    this.tokens.push(new Token('BOOLEAN', false, this.line, startCol + pos));
                } else if (val === 'None' || val === 'null') {
                    this.tokens.push(new Token('NONE', null, this.line, startCol + pos));
                } else {
                    const num = Number(val);
                    if (!isNaN(num) && val.trim() !== '') {
                        this.tokens.push(new Token('NUMBER', num, this.line, startCol + pos));
                    } else {
                        this.tokens.push(new Token('IDENTIFIER', val, this.line, startCol + pos));
                    }
                }
                pos += val.length;
            } else {
                throw new Error(`Unexpected character '${c}' at line ${this.line}`);
            }
        }
    }
}

export class Parser {
    private pos: number = 0;

    constructor(private tokens: Token[]) {}

    private peek(): Token {
        if (this.pos < this.tokens.length) {
            return this.tokens[this.pos];
        }
        return this.tokens[this.tokens.length - 1];
    }

    private consume(expectedType?: string): Token {
        const tok = this.peek();
        if (expectedType && tok.type !== expectedType) {
            throw new Error(`Expected ${expectedType}, got ${tok.type} at line ${tok.line}`);
        }
        this.pos++;
        return tok;
    }

    private _skipNewlines() {
        while (this.peek().type === 'NEWLINE') {
            this.consume('NEWLINE');
        }
    }

    public parse(): any {
        this._skipNewlines();
        const tok1 = this.peek();
        if (tok1.type === '[') {
            return this.parseList();
        }
        return this.parseDict();
    }

    private _skipKeyAnnotation() {
        if (this.peek().type === '[') {
            if (this.pos + 1 < this.tokens.length) {
                const nextTok = this.tokens[this.pos + 1];
                if (nextTok.type === ']') return;
                if (nextTok.type === 'NUMBER') {
                    if (this.pos + 2 < this.tokens.length && this.tokens[this.pos + 2].type === ']') {
                        return;
                    }
                }
                
                this.consume('[');
                while (![']', 'NEWLINE', 'EOF'].includes(this.peek().type)) {
                    this.consume();
                }
                if (this.peek().type === ']') {
                    this.consume(']');
                }
            }
        }
    }

    private parseDict(endTokens: string[] = ['EOF', 'DEDENT', ')', ']']): any {
        const res: any = {};
        while (!endTokens.includes(this.peek().type)) {
            this._skipNewlines();
            if (endTokens.includes(this.peek().type)) break;

            const keyTok = this.peek();
            if (!['IDENTIFIER', 'NUMBER', 'STRING'].includes(keyTok.type)) {
                throw new Error(`Expected IDENTIFIER, NUMBER or STRING as dict key, got ${keyTok.type} at line ${keyTok.line}`);
            }
            this.consume();
            const key = String(keyTok.value);

            this._skipKeyAnnotation();

            let arrayDepth = 0;
            let isKeyedTabular = false;
            while (true) {
                if (this.peek().type === '[') {
                    if (this.pos + 1 < this.tokens.length && this.tokens[this.pos + 1].type === ']') {
                        this.consume('[');
                        this.consume(']');
                        arrayDepth++;
                    } else if (this.pos + 1 < this.tokens.length && this.tokens[this.pos + 1].type === 'NUMBER') {
                        if (this.pos + 2 < this.tokens.length && this.tokens[this.pos + 2].type === ']') {
                            this.consume('[');
                            const num = Number(this.consume('NUMBER').value);
                            this.consume(']');
                            arrayDepth += num;
                        } else {
                            break;
                        }
                    } else {
                        break;
                    }
                } else if (this.peek().type === '{') {
                    if (this.pos + 1 < this.tokens.length && this.tokens[this.pos + 1].type === '}') {
                        this.consume('{');
                        this.consume('}');
                        isKeyedTabular = true;
                    } else if (this.pos + 1 < this.tokens.length && this.tokens[this.pos + 1].type === '[') {
                        if (this.pos + 2 < this.tokens.length && this.tokens[this.pos + 2].type === ']') {
                            if (this.pos + 3 < this.tokens.length && this.tokens[this.pos + 3].type === '}') {
                                this.consume('{');
                                this.consume('[');
                                this.consume(']');
                                this.consume('}');
                                isKeyedTabular = true;
                                arrayDepth++;
                            } else break;
                        } else break;
                    } else break;
                } else {
                    break;
                }
            }

            if (this.peek().type === 'NEWLINE') {
                if (this.pos + 1 < this.tokens.length && this.tokens[this.pos + 1].type === 'INDENT') {
                    res[key] = this.parseTabularBlock(arrayDepth, isKeyedTabular);
                    continue;
                } else {
                    if (arrayDepth > 0) {
                        res[key] = [];
                    } else {
                        res[key] = {};
                    }
                    continue;
                }
            }

            if (this.peek().type === '=') {
                this.consume('=');
                res[key] = this.parseValue();
            } else if (this.peek().type === '[') {
                res[key] = this.parseList();
            } else if (this.peek().type === '(') {
                res[key] = this.parseInlineGroup();
            } else {
                res[key] = this.parseValue();
            }
        }
        return res;
    }

    private parseTabularBlock(arrayDepth: number, isKeyedTabular: boolean = false): any {
        this.consume('NEWLINE');
        this.consume('INDENT');

        if (arrayDepth >= 2 || (isKeyedTabular && arrayDepth >= 1)) {
            const res: any = isKeyedTabular ? {} : [];
            let currentSlice: any[] = [];
            
            let newlineCount = 0;
            while (this.peek().type === 'NEWLINE') {
                this.consume('NEWLINE');
                newlineCount++;
            }
            
            while (!['DEDENT', 'EOF'].includes(this.peek().type)) {
                if (this.peek().type === 'DEDENT') break;
                
                if (arrayDepth >= 3 && newlineCount >= 2 && currentSlice.length > 0) {
                    if (!isKeyedTabular) {
                        res.push(currentSlice);
                    }
                    currentSlice = [];
                }
                
                let dictKey: string | null = null;
                if (isKeyedTabular) {
                    const keyTok = this.peek();
                    if (!['IDENTIFIER', 'NUMBER', 'STRING'].includes(keyTok.type)) {
                        throw new Error(`Expected dict key for keyed matrix row, got ${keyTok.type} at line ${keyTok.line}`);
                    }
                    this.consume();
                    dictKey = String(keyTok.value);
                }
                
                const row = this.parseTupleRow();
                
                if (arrayDepth >= 3) {
                    currentSlice.push(row);
                } else {
                    if (isKeyedTabular) {
                        res[dictKey!] = row;
                    } else {
                        res.push(row);
                    }
                }
                
                newlineCount = 0;
                while (this.peek().type === 'NEWLINE') {
                    this.consume('NEWLINE');
                    newlineCount++;
                }
            }
            
            if (arrayDepth >= 3 && currentSlice.length > 0) {
                res.push(currentSlice);
            }
            
            this.consume('DEDENT');
            return res;
        }

        const headers: any[] = [];
        while (!['NEWLINE', 'EOF'].includes(this.peek().type)) {
            const keyTok = this.peek();
            if (!['IDENTIFIER', 'NUMBER', 'STRING'].includes(keyTok.type)) {
                throw new Error(`Expected header key, got ${keyTok.type} at line ${keyTok.line}`);
            }
            this.consume();
            const colKey = String(keyTok.value);

            this._skipKeyAnnotation();

            while (this.peek().type === '[') {
                if (this.pos + 1 < this.tokens.length && this.tokens[this.pos + 1].type === ']') {
                    this.consume('[');
                    this.consume(']');
                } else break;
            }

            while (this.peek().type === '(') {
                if (this.pos + + 1 < this.tokens.length && this.tokens[this.pos + 1].type === ')') {
                    this.consume('(');
                    this.consume(')');
                } else break;
            }

            if (this.peek().type === '(') {
                this.consume('(');
                const subKeys: string[] = [];
                while (![')', 'NEWLINE', 'EOF'].includes(this.peek().type)) {
                    const subKeyTok = this.peek();
                    if (!['IDENTIFIER', 'NUMBER', 'STRING'].includes(subKeyTok.type)) {
                        throw new Error(`Expected sub header key, got ${subKeyTok.type}`);
                    }
                    this.consume();
                    subKeys.push(String(subKeyTok.value));
                    this._skipKeyAnnotation();
                }
                this.consume(')');
                headers.push([colKey, subKeys]);
            } else {
                headers.push(colKey);
            }
        }

        this.consume('NEWLINE');

        const res: any = isKeyedTabular ? {} : [];
        while (!['DEDENT', 'EOF'].includes(this.peek().type)) {
            this._skipNewlines();
            if (this.peek().type === 'DEDENT') break;

            let dictKey: string | null = null;
            if (isKeyedTabular) {
                const keyTok = this.peek();
                if (!['IDENTIFIER', 'NUMBER', 'STRING'].includes(keyTok.type)) {
                    throw new Error(`Expected dict key for keyed tabular row, got ${keyTok.type} at line ${keyTok.line}`);
                }
                this.consume();
                dictKey = String(keyTok.value);
            }

            const rowDict: any = {};
            for (const header of headers) {
                if (Array.isArray(header)) {
                    const mainK = header[0];
                    const subK = header[1];
                    const val = this.parseValue();
                    if (Array.isArray(val)) {
                        const subDict: any = {};
                        for (let i = 0; i < subK.length; i++) {
                            subDict[subK[i]] = val[i];
                        }
                        rowDict[mainK] = subDict;
                    } else if (typeof val === 'object' && val !== null) {
                        const subDict: any = { ...val };
                        for (let i = 0; i < subK.length; i++) {
                            const strI = String(i);
                            if (strI in subDict) {
                                subDict[subK[i]] = subDict[strI];
                                delete subDict[strI];
                            }
                        }
                        rowDict[mainK] = subDict;
                    } else {
                        throw new Error(`Expected array or dict for ${mainK} at line ${this.peek().line}`);
                    }
                } else {
                    rowDict[header] = this.parseValue();
                }
            }

            while (!['NEWLINE', 'EOF', 'DEDENT'].includes(this.peek().type)) {
                const keyTok = this.peek();
                if (!['IDENTIFIER', 'NUMBER', 'STRING'].includes(keyTok.type)) {
                    throw new Error(`Expected dynamic property key, got ${keyTok.type} at line ${keyTok.line}`);
                }
                this.consume();
                const key = String(keyTok.value);

                this._skipKeyAnnotation();

                if (this.peek().type === '=') {
                    this.consume('=');
                    rowDict[key] = this.parseValue();
                } else {
                    throw new Error(`Expected '=' after dynamic property '${key}' at line ${keyTok.line}`);
                }
            }

            if (isKeyedTabular) {
                res[dictKey!] = rowDict;
            } else {
                res.push(rowDict);
            }
        }

        this.consume('DEDENT');
        if (isKeyedTabular) return res;
        return arrayDepth > 0 ? res : res[0];
    }

    private parseList(): any[] {
        this.consume('[');
        const res: any[] = [];
        while (this.peek().type !== ']') {
            while (['NEWLINE', 'INDENT', 'DEDENT'].includes(this.peek().type)) {
                this.consume();
            }
            if (this.peek().type === ']') break;
            res.push(this.parseValue());
        }
        this.consume(']');
        return res;
    }

    private parseTupleRow(): any[] {
        const row: any[] = [];
        while (!['NEWLINE', 'EOF', 'DEDENT'].includes(this.peek().type)) {
            row.push(this.parseValue());
            if (this.peek().type === ',') {
                this.consume(',');
            }
        }
        return row;
    }

    private parseInlineGroup(): any {
        this.consume('(');
        const res: any = {};
        let posIdx = 0;
        let hasKwargs = false;

        while (this.peek().type !== ')') {
            while (['NEWLINE', 'INDENT', 'DEDENT'].includes(this.peek().type)) {
                this.consume();
            }
            if (this.peek().type === ')') break;

            let temp = this.pos;
            const tok = this.tokens[temp];
            let isKwarg = false;

            if (['IDENTIFIER', 'NUMBER', 'STRING'].includes(tok.type)) {
                temp++;
                while (temp < this.tokens.length && this.tokens[temp].type === '[') {
                    temp++;
                    if (temp < this.tokens.length && this.tokens[temp].type === ']') {
                        temp++;
                    } else {
                        break;
                    }
                }
                if (temp < this.tokens.length && this.tokens[temp].type === '=') {
                    isKwarg = true;
                }
            }

            if (isKwarg) {
                hasKwargs = true;
                const keyTok = this.peek();
                this.consume();
                const key = String(keyTok.value);
                this._skipKeyAnnotation();
                this.consume('=');
                res[key] = this.parseValue();
            } else {
                const val = this.parseValue();
                res[String(posIdx)] = val;
                posIdx++;
            }
        }
        this.consume(')');

        if (!hasKwargs) {
            const arr = [];
            for (let i = 0; i < posIdx; i++) arr.push(res[String(i)]);
            return arr;
        }

        return res;
    }

    private parseValue(): any {
        const tok = this.peek();
        if (tok.type === '[') return this.parseList();
        if (tok.type === '(') return this.parseInlineGroup();
        if (['STRING', 'NUMBER', 'BOOLEAN', 'NONE', 'IDENTIFIER'].includes(tok.type)) {
            this.consume();
            return tok.value;
        }
        throw new Error(`Unexpected token ${tok.type} in parseValue at line ${tok.line}`);
    }
}

export function parse(text: string): any {
    const lexer = new Lexer(text);
    const tokens = lexer.tokenize();
    const parser = new Parser(tokens);
    return parser.parse();
}
